peas/tests/serial_handlers/protocol_arduino_simulator.py

```diff
@@ -0,0 +1,422 @@
+    will generate another json output line,
+    at around 1000 bytes per second (near 9600 buad). The bytes  (actually chunked up in batches, so not emitting a single byte
+    every 1/960th of a second or so).
+    """
+
+    def __init__(self, relay_queue, json_queue, stop, logger):
+        """
+        Args:
+            relay_queue: The queue.Queue instance from which replay command bytes are read
+                and acted upon. Elements are of type bytes.
+            json_queue: The queue.Queue instance to which json messages (serialized to bytes)
+                are written at ~9600 baud. Elements are of type bytes.
+            stop: a threading.Event which is checked to see if run should stop executing.
+            logger: the Python logger to use for reporting messages.
+        """
+        self.relay_queue = relay_queue
+        self.json_queue = json_queue
+        self.stop = stop
+        self.logger = logger
+        # Time between producing messages.
+        self.message_delta = datetime.timedelta(seconds=2)
+        self.next_message_time = None
+        # Size of a chunk of bytes.
+        self.chunk_size = 20
+        # Interval between outputing chunks of bytes.
+        chunks_per_second = 1000.0 / self.chunk_size
+        chunk_interval = 1.0 / chunks_per_second
+        self.logger.debug('chunks_per_second={}   chunk_interval={}',
+                          chunks_per_second, chunk_interval)
+        self.chunk_delta = datetime.timedelta(seconds=chunk_interval)
+        self.next_chunk_time = None
+        self.pending_json_bytes = bytearray()
+        self.pending_relay_bytes = bytearray()
+        self.start_time = datetime.datetime.now()
+        self.report_num = 0
+        # TODO(jamessynge): Come up with a way to select which board is emulated. For now just
+        # one board is supported.
+        # TODO(jamessynge): Update camera_board message soon when camera board sketch is updated.
+        self.camera_board = json.loads("""
+            {
+                "name":"camera_board",
+                "millis":1507497677,
+                "report_num":7835,
+                "inputs":6,
+                "camera_00":1,
+                "camera_01":1,
+                "accelerometer": {"x":-7.02, "y":6.95, "z":1.70, "o": 6},
+                "humidity":59.60, "temp_00":12.50
+            }
+            """)
+        # TODO(jamessynge): Update telemetry_board message to reflect latest changes.
+        self.telemetry_board = json.loads("""
+            {
+                "name":"telemetry_board",
+                "millis":155581294,
+                "report_num":4801,
+                "ver":"2017-09-23",
+                "power": {"computer":1, "fan":1, "mount":1, "cameras":1, "weather":1, "main":1},
+                "current": {"main":387,"fan":28,"mount":34,"cameras":27},
+                "amps": {"main":1083.60,"fan":50.40,"mount":61.20,"cameras":27.00},
```

Comment:
I realize you are mocking what is there now, but I thought we wanted to drop the amps from the arduino code.

I guess I say that because the worry here, and with comment above, is that this test just gets out of line with what we are actually testing for. Not sure there is any real way around that.

Reply:
Agreed, I don't see a good way around that... besides an awesome arduino simulator, with mock sensors attached!

Reply:
I dream about building a whole unit and storing it in a lab at Subaru and then hooking the testing up via Jenkins so we can run real automated hardware tests. But that's a long way off.